instagram_search_by.py

```diff
@@ -0,0 +1,154 @@
+
+    def _process_response(self, resp):
+        """ Extract fresh posts from the Instagram api response object.
+
+        Args:
+            resp (Response)
+
+        Returns:
+            signals (list(Signal)): The list of signals to notify, each of
+                which corresponds to a fresh Instagram post.
+            paging (bool): Denotes whether or not paging requests are
+                necessary.
+
+        """
+        signals = []
+        resp = resp.json()
+
+        pagination = resp.get('pagination', [])
+        paging = self._check_paging(pagination)
+
+        fresh_posts = posts = resp.get('data', [])
+        if len(posts) > 0:
+            self.update_freshness(posts)
+            fresh_posts = self.find_fresh_posts(posts)
+
+        signals = [Signal(p) for p in fresh_posts]
+        self._logger.info("Created {0} new Instagram signals.".format(
+            len(signals)))
+
+        return signals, paging
+
+    def _get_post_id(self, post):
+        return getattr(post, 'id', None)
+
+    def _retry(self, resp, paging):
+        resp_error_type = resp.json().get('meta', {}).get('error_type')
+        if resp.status_code == 400 and \
+           (resp_error_type == 'APINotAllowedError' or
+            resp_error_type == 'APINotFoundError'):
+            # this is a private user, skip to next query.
+            self._logger.debug("Skipping private user: {}".format(
+                self.current_query))
+            self._increment_idx()
+        else:
+            super()._retry(resp, paging)
+
+    def _check_paging(self, pagination):
+        if 'next_url' in pagination:
+            self.url = pagination['next_url']
+            return True
+        else:
+            return False
+
+    def _process_query(self, query):
+        """ Queries need to be converted from username to id.
+
+        Instagram api queries by user need a user id as a parameter.
+        This block asks has a parameter for usernames so on start,
+        requests to the instagram api need to be made to convert
+        username to id.
+
+        """
+        # If there is a cached id for this user, return it
+        # and skip the request
+        _id = self.persistence.load(query.lower())
+        if _id is not None:
+            return _id
+        
+        resource_url = self._construct_resource_url(query)
+        
+        resp = self._make_request(resource_url)
+        if resp is None:
+            # try again if the response was bad.
+            self._logger.debug("Attempting immediate retry to get id"
+                               " for: {0}".format(query))
+            resp = self._make_request(resource_url)
+            if resp is None:
+                # if the response is still bad, give up.
+                return
+
+        resources = resp.json().get('data', [])
+        return self._extract_resource_id(resources, query)
+
+    def _construct_resource_url(self, query):
+        return self.RESOURCE_URL_FORMAT.format(
```

Comment:
Ya. Unintentional omissions. The latter is easily fixed, the former has some grey area. I'll have to mull it over.